projects/sandbox/datagen/datagen/utils/priors.py

```diff
@@ -1,109 +1,131 @@
-            unit=r"$M_{\odot}",
-        )
-        prior_dict["luminosity_distance"] = UniformComovingVolume(
-            name="luminosity_distance", minimum=100, maximum=15000, unit="Mpc"
-        )
-        prior_dict["dec"] = Cosine(name="dec")
-        prior_dict["ra"] = Uniform(
-            name="ra", minimum=0, maximum=2 * np.pi, boundary="periodic"
-        )
-        prior_dict["theta_jn"] = 0
-        prior_dict["psi"] = 0
-        prior_dict["phase"] = 0
-        prior_dict["a_1"] = Uniform(name="a_1", minimum=0, maximum=0.998)
-        prior_dict["a_2"] = Uniform(name="a_2", minimum=0, maximum=0.998)
-        prior_dict["tilt_1"] = Sine(name="tilt_1", unit="rad")
-        prior_dict["tilt_2"] = Sine(name="tilt_2", unit="rad")
-        prior_dict["phi_12"] = Uniform(
-            name="phi_12", minimum=0, maximum=2 * np.pi, boundary="periodic"
-        )
-        prior_dict["phi_jl"] = 0
-
-        return prior_dict
-
-
-extrinsic_params = ExtrinsicParams()
-nonspin_bbh = NonSpinBBH()
-end_o3_ratesandpops = EndO3RatesAndPops()
+from bilby.gw.prior import (
+    BBHPriorDict,
+    UniformComovingVolume,
+    UniformSourceFrame,
+)
+
+# Unit names
+msun = r"$M_{\odot}$"
+mpc = "Mpc"
+rad = "rad"
+
+
+def pdf_from_events(
+    param_values: Sequence[float],
+    grid_size: int = 100,
+    spacing: str = "lin",
+):
+    param_min = np.min(param_values)
+    param_max = np.max(param_values)
+    if spacing == "lin":
+        bins = np.linspace(param_min, param_max, grid_size + 1)
+        grid = (bins[:-1] + bins[1:]) / 2
+    elif spacing == "log":
+        min_exp = np.log10(param_min)
+        max_exp = np.log10(param_max)
+        bins = np.logspace(min_exp, max_exp, grid_size + 1)
+        grid = np.sqrt(bins[:-1] * bins[1:])
+    else:
+        raise ValueError("Spacing must be either 'lin' or 'log'")
+
+    pdf, _ = np.histogram(param_values, bins, density=True)
+
+    return grid, pdf
+
+
+def read_priors_from_file(event_file: Path, *parameters: str) -> BBHPriorDict:
+    prior = BBHPriorDict()
+    with h5py.File(event_file, "r") as f:
+        events = f["events"]
+        events = events[(events["mass_1"] > 5) & (events["mass_2"] > 5)]
+        field_names = parameters or events.dtype.names
+        for name in field_names:
+            # We'll specify mass_2 via mass_ratio at sampling time
+            if name == "mass_2":
+                continue
+            grid, pdf = pdf_from_events(events[name])
+            prior[name] = Interped(grid, pdf, np.min(grid), np.max(grid))
+
+    return prior
+
+
+def assign_prior_names(prior: BBHPriorDict) -> BBHPriorDict:
+    for key in prior.keys():
+        if hasattr(prior[key], "name"):
+            print("check")
+            prior[key].name = key
+    return prior
+
+
+def uniform_extrinsic():
+    prior = BBHPriorDict()
+    prior["dec"] = Cosine()
+    prior["ra"] = Uniform(0, 2 * np.pi)
+    prior["theta_jn"] = 0
+    prior["phase"] = 0
+    prior = assign_prior_names(prior)
```

Comment:
I really like this functionality, I've always thought the naming and key assignment was redundant. Out of curiosity, what happens if the parameters don't have names?

Reply:
Agreed. I actually moved away from using this function in the latest commit because most of the priors don't need names. For the two that do, I've hardcoded the names in.

Reply:
I *think* the names are used for bilbys plotting functionality (i.e. they are optional). Will, which priors need names, and do you know why?

Reply:
Generally, anything that inherits from bilby's `Cosmological` class requires the name to be specified, and to be in `["redshift", "luminosity_distance", "comoving_distance"]`. For us, that means we need to specify names at initialization for `UniformComovingVolume` and `UniformSourceFrame`.